### ISSUE TITLE
[Issue]: GBA.emu not using all of the available CPU time

### ISSUE THREAD

**Issue by Code-Otto** [Rakashazi/emu-ex-plus-alpha#1303]
### Describe the bug

Recently gave a try upgrading from the old and (subjectively) trusty GbaEmu-1.5.27-beta1 that used to be hosted at your website to the very latest f41938e3b6cf48868e8e43b098fcae54e146b034 nightly on my Android 7.1.2 device and I'm facing a performance issue where the CPU stays at lowest frequency despite the emulator not yet running at full speed

The phone is a 'BQ Aquaris U Plus' I have been able to reproduce the issue both under its stock firmware and LineageOS 14.1 which is my daily driver

Here's a table with the CPU usage of both emulator versions under different conditions:
```
             |  CPU locked at | Automatic CPU freq    | CPU locked at
             |  lowest freq   | (Android default mode)| max freq
------------------------------------------------------------------------------------
             | Game lags      | Game lags             | Game at full speed
Latest       | ~85% CPU usage | ~85% CPU usage        | ~65% CPU usage
             | 960MHz (locked)| 960MHz (auto)         | 1401MHz (locked)
------------------------------------------------------------------------------------
             | Game lags      | Game at full speed    | Game at full speed
1.5.27-beta1 | 100% CPU usage | ~75% CPU usage        | ~50% CPU usage
             | 960MHz (locked)| 1209MHz (auto)        | 1401MHz (locked)
------------------------------------------------------------------------------------
```
(CPU usage measured on the com.explusalpha.GbaEmu process)

I believe because of not pushing the CPU to 100% usage, the Linux CPU frequency scaling governor never acknowledges the need of setting a higher frequency keeping the emulator below its true performance needs

Tried changing the following settings without it causing any improvement:
 - Skip Late Frames On/Off
 - Performance Mode Normal/Sustained
 - Frame Rate
 - GPU Copy Mode

Also reproduced the issue in an older Samsung Android KitKat device, however a friend's powerful Android 13 device had no noticeable speed issues

Please tell me about any other helpful information and I will try to collect it 

Thanks for making this awesome set of emulators, I've used them for many years even back to the days of the 'Free!' versions at Play Store and it has always proven a reliable and pleasing experience. You know someone puts a lot of love and care in making their app when it is 2023 yet sdkVersion is still at 9 :heart: (Yeah if you ever wondered there are we nerdy folks out there finding that extremely cool)

### Steps to reproduce

 - Install latest GBA.emu in a oldish Android phone
 - Use a tool like this https://forum.xda-developers.com/t/app-cpufloat-march27-floating-cpu-gpu-temperatures-deep-sleep-network-monitor.3204420/ to monitor CPU frequencies
 - Run a demanding game (I'm using Iridion 2)
 
If you get the same fate as my devices, CPU will stay at lower frequencies despite emulation not yet running full speed

### Expected behavior

The emulator pushes to 100% CPU usage as needed, telling the CPU frequency governor more performance is required

**Comment by Code-Otto** [comment]
The 'Reduce Compositor Lag' option didn't yield any changes either unfortunately. The issues I'm facing aren't just eventual frame drops but a consistently low frame rate and choppy sound

I've however been further investigating the issue and found what may be the culprit
The Qualcomm MSM8937 powering this device features a 8-core ARM CPU, 4 of them being 'low-power' cores with reduced performance
Monitoring the emulator process I've noticed the system always runs it off the low-power cores, which explains the poor performance. For some reason the ancient 1.5.27-beta1 version gets to run on the faster cores instead explaining the good results

If I manually shutdown the low-power cores via sysfs to force the emulator into the faster cores all the performance issues go away and I get the emulator to a playable frame rate

Does Android allow app processes to set affinity to a certain CPU core(s)? If so, making it user-configurable could help working a way out of the issue

**Comment by Rakashazi** [comment]
I searched through the APIs and found [PerformanceHintManager](https://developer.android.com/reference/android/os/PerformanceHintManager) that can influence how threads are scheduled on CPU but unfortunately it's only available on Android 12+. I'll keep looking and see if I can find something for older Android versions.

**Comment by Code-Otto** [comment]
Did the following experiment and it works:
```# run-as com.explusalpha.GbaEmu taskset -a -p 0F `pidof com.explusalpha.GbaEmu` ``` (0F masks away the low-power cores in my device)
So it seems within the privileges of the app the `sched_setaffinity` and `sched_getaffinity` syscalls are permitted

Here's the sources of toybox's taskset for reference (BSD license): https://android.googlesource.com/platform/external/toybox/+/refs/heads/master/toys/other/taskset.c

**Comment by Code-Otto** [comment]
I've been working out some logic to query CPU details via sysfs in Android within standard app permissions. I set it up as a shell script for ease however Java and native code should be able to perform the same in an app

```sh
#!/system/bin/sh

# Queries CPU information from sysfs in Android w/o root permissions
# Code-Otto 2023

# There's no access to list the /sys/devices/system/cpu/ directory in
# Android but each cpuX subdir is readable and may be manually queried
i=0
while [ -d /sys/devices/system/cpu/cpu$i ]
do
	echo "#### CPU $i ####"
	
	### Check if the CPU is online
	if [ -f /sys/devices/system/cpu/cpu$i/online ]
	then
	online=`cat /sys/devices/system/cpu/cpu$i/online`
	else
	# Core can't be shut down, ie CPU0 in some x86 kernels
	online=1
	fi
	
	if [ "$online" -eq 1 ]
	then
		echo "Status: Online"
	else
		echo "Status: Offline"
	fi
	
	
	# Check the available frequencies and CPU groups
	if [ -d /sys/devices/system/cpu/cpu$i/cpufreq ]
	then
		echo -n "Related CPUs: "
		cat /sys/devices/system/cpu/cpu$i/cpufreq/related_cpus
		echo -n "Max. Freq (hardware): "
		cat /sys/devices/system/cpu/cpu$i/cpufreq/cpuinfo_max_freq
		# scaling_max_freq is unreadable in some devices
		echo -n "Max. Freq (governor): "
		cat /sys/devices/system/cpu/cpu$i/cpufreq/scaling_max_freq
		echo -n "Min. Freq (hardware): "
		cat /sys/devices/system/cpu/cpu$i/cpufreq/cpuinfo_min_freq
		echo -n "Min. Freq (governor): "
		cat /sys/devices/system/cpu/cpu$i/cpufreq/scaling_min_freq
		
	else
		if [ "$online" -eq 1 ]
		then
			echo "CPU frequencies unavailable (unknown error)"
		else
			# In ARM systems cpuX/cpufreq isn't available while the core
			# is offline
			echo "CPU frequencies unavailable (core is offline)"
		fi
	fi
	
	i=$((i+1))
done

echo "- Total CPUs: $i"
```

Tested in an Android 7.1.2 BQ, a 4.4.4 Samsung, a 4.1.2 LG and a 2.3.7 LG as well as a x86 laptop

Here's the output from my BQ (4,5,6,7 are the low-power cores)
```
#### CPU 0 ####
Status: Online
Related CPUs: 0 1 2 3
Max. Freq (hardware): 1401000
Max. Freq (governor): 1401000
Min. Freq (hardware): 960000
Min. Freq (governor): 960000
#### CPU 1 ####
Status: Online
Related CPUs: 0 1 2 3
Max. Freq (hardware): 1401000
Max. Freq (governor): 1401000
Min. Freq (hardware): 960000
Min. Freq (governor): 960000
#### CPU 2 ####
Status: Offline
CPU frequencies unavailable (Core is offline)
#### CPU 3 ####
Status: Offline
CPU frequencies unavailable (Core is offline)
#### CPU 4 ####
Status: Online
Related CPUs: 4 5 6 7
Max. Freq (hardware): 1094400
Max. Freq (governor): 1094400
Min. Freq (hardware): 768000
Min. Freq (governor): 768000
#### CPU 5 ####
Status: Online
Related CPUs: 4 5 6 7
Max. Freq (hardware): 1094400
Max. Freq (governor): 1094400
Min. Freq (hardware): 768000
Min. Freq (governor): 768000
#### CPU 6 ####
Status: Online
Related CPUs: 4 5 6 7
Max. Freq (hardware): 1094400
Max. Freq (governor): 1094400
Min. Freq (hardware): 768000
Min. Freq (governor): 768000
#### CPU 7 ####
Status: Online
Related CPUs: 4 5 6 7
Max. Freq (hardware): 1094400
Max. Freq (governor): 1094400
Min. Freq (hardware): 768000
Min. Freq (governor): 768000
- Total CPUs: 8
```

Some notes:
 - `cpufreq/scaling_max_freq` was unreadable in the 4.x.x devices, `cpufreq/cpuinfo_max_freq` was still OK
 - `cpufreq/related_cpus` may be useful for grouping CPU cores in the user interface
 - `cpufreq/` is not available in ARM systems while the core is offline. As a workaround it may be possible to spawn a thread affine to that single offline CPU as an attempt to temporarily force its activation and be able to peek at the details

Hope these details can be of help. I don't have access to any of the other platforms targeted by your emulator so can't offer any insights for them unfortunately

**Comment by Rakashazi** [comment]
Thanks for all the additional info. I added a menu under Options -> System -> Override CPU Affinity that lists all CPUs along with their frequencies (from cpuinfo_max_freq) to help identify which cores are big/little. By toggling CPUs to "on" they'll be added to the affinity mask and having them all "off" is the same as all CPUs active so it's easier to test one CPU vs all active. Let me know if it's working properly on your devices.

**Comment by Code-Otto** [comment]
The new menu works like a charm, thank you!

It would be useful for all cores to appear even while they are offline
![screen2](https://user-images.githubusercontent.com/129909770/231278832-569a9b1f-2d06-4542-903d-3c4d962e3074.png)
Since Android (at least in my device) picks different cores every time the app is started/switched to, it takes some trial and error to be able to fully configure the entire affinity mask. I feel that may get confusing for non-technical users who see the established configuration arbitrarily change on each launch

This allows emulation to work at a playable frame rate on my device without the burden of messing with a root adb shell every playing session :smiley:, thanks for taking the time to implement this :heart: 

However something still feels wrong. This mitigates the problem but the issue I described at the first post is still there
Latest Gba.Emu can run at full frame rate in my device if I manually force the processor into a higher frequency but the emulator on its own does not push the CPU into such state. Feels as if some time was being wasted waiting for threads or locked resources

I'm however very far from a wise and knowledgeable programmer, much less aware of inner details on this codebase you've crafted through the years. Best I feel I can do is trying different versions between 1.5.27 and latest to find out the exact one this regression was introduced at (if such a single one exists, instead of multiple issues cascading) so I can point you at it

Is there anywhere I could get hold of prebuilt old versions of the emulator to test them out? Will try getting a build environment set up otherwise but that may take me some time
Also, is there any way to measure the FPS or how many frames are being skipped?

**Comment by Rakashazi** [comment]
Good to hear it worked, I'll update the menu to show offline cores in the next commit.

The CPU frequency issue is most likely the CPU governor not ramping up fast enough to respond to the current load. I don't think it's related to locked resources but there's a chance the sync fence used between frames (added in 1.5.47) is creating a stall so try setting Options -> Video -> Images Buffer -> 2, which won't use any fences. It's possible driver bugs will make the fence wait longer than it's supposed to so this will rule that out.

I also used to explicitly set the process priority (aka nice value) to -6 and provide an option to set other values but didn't see much of any difference in testing various devices so it was eventually removed. I wonder if the priority value was affecting the governor on that device in some way.

Here's an APK of 1.5.46 which is the last version with the process priority option so it's a good version to start testing with. Let me know how it compares to the latest version and 1.5.27.

https://drive.google.com/file/d/1GtXPdMxW9yEvMlwHuw1jZE9omzQaS6MY/view

We can try out other versions once we have a better idea of how this one behaves. There's no in-app frame time stats but I'm looking into adding some for a future update. For now your best bet is to use a game with a scrolling background so the frame drops are easier to see.

**Comment by Code-Otto** [comment]
Tested changing the 'Images Buffer' option and it caused no changes either, unfortunately

Gave 1.5.46 a try and it experiences quite the same issue as the latest versions do. Changing the process priority from the UI yields no changes neither did changing the performance mode, which defaults to 'Sustained' in this version

Found out yesterday about scrcpy (https://github.com/Genymobile/scrcpy) and it does an awesome job at smoothly streaming the device screen via ADB, so great it can make good grabs of the frame rate issues I'm experiencing. Here's a comparison video between 1.5.27-beta1, 1.5.46 and latest with CPU affinity configured for your viewing (dis)pleasure :smiley: 

**1.5.27-beta1:**
The weapon selection screen is quite heavy on emulation resources however ingame the CPU stays at a midpoint around 1200MHz. The frame rate feels high and there's almost no drops

https://user-images.githubusercontent.com/129909770/231588837-6db837b2-9cdd-4d13-bb00-2185cd21b18e.mp4


**1.5.46:**
Performs quite the same as the latest versions, frame rate feels halved. Stays in the low-power CPU cores (4-7). Notice how while enabling fast-forward the process moves away from the low-power cores but gets back in them as soon as disabled

https://user-images.githubusercontent.com/129909770/231588926-89db6956-9f0a-43f9-bce8-b5308bd4e395.mp4

**Commit 1b03458b843e2ef59a7819b2b247f0fd43ffbee1 (latest as of this post):**
Setting affinity to the performance cores makes the frame rate much smoother but still feels slightly off. Notice the low CPU frequencies even while the emulator clearly lags, specially while 'Skip late frames' is disabled

https://user-images.githubusercontent.com/129909770/231593662-3479f5b0-cbab-4107-9090-6a23394229c2.mp4

As a fun fact background apps pushing up the CPU frequency momentarily make the emulator run smooth, for example if a e-mail notification pops up

**Comment by Rakashazi** [comment]
Thanks for the report, good to know the process priority wasn't affecting anything and that option wasn't needed. Android should manage the priority itself so there was no reason to mess with it.

Please try out the new commit and this time keep Image Buffers at 1 and Reduce Compositor Lag on. This setup no longer uses GPU sync fences so we'll see if it affects performance. Here are the other changes of note to test out:

- Offline CPUs should now show up in the affinity menu
- There's a new option under System to enable a thread executing no-op instructions, this combined with CPU affinity may allow the CPU core running the emulation to stay at a high frequency.

It's funny you mentioned your email notification increasing the CPU frequency because while testing these changes I discovered my bank's app was randomly waking up and using 100% of one core in the background lol. As a result, I had to put it into battery restricted mode to make sure it wasn't causing interference.

**Comment by Code-Otto** [comment]
To begin with, the revised CPU affinity menu works great. Does its job and it's now easier to configure :heart: 

Gave this new version a try and found out the following two options yielding some notable performance changes with all default settings as you recommended and CPU affinity set up:
- First, the classical scenarios we've already tested a lot:
No-op thread OFF + Skip late frames ON -> Low frame rate
No-op thread OFF + Skip late frames OFF -> Game runs slowly
- With the no-op thread the CPU is forced into max frequency as expected and brings in some new results:
No-op thread ON + Skip late frames ON -> Performance feels better and full frame rate is reached for some short spans however there's still frame dropping
No-op thread ON + Skip late frames OFF -> Runs smoothly almost like 1.5.27-beta1! :smiley: 

Captured as well a video going through said different configurations, starting from defaults:
Notice how after setting No-op thread ON + Skip late frames ON and closing the menu, the game runs smoothly for around a second then starts dropping frames again

https://user-images.githubusercontent.com/129909770/231805589-bb2a0136-0f21-48d5-b4c3-93df9e018276.mp4

Visually estimating frame rates feels somewhat inaccurate and sometimes I wonder if this entire issue is just some paranoia-induced bias on my senses :cold_sweat: 

I've as well tested changing the 'Image Buffers' and 'Reduce Compositor Lag' options alongside the no-op thread setting without them yielding any apparent change at all

As an interesting note, if I do not configure any CPU affinity and enable the no-op thread it does not escalates out of the low-power cores at all. Found out after some experimenting any process with a niceness of 16 or above won't escalate to the performance cores regardless of CPU time (as configured in `/proc/sys/kernel/sched_upmigrate_min_nice`)
However high priorities appear not to cause the opposite effect of forcing performance cores not either there seems to be such a kernel tunable for it (At this point I feel I should dedicate some time into learning how the kernel scheduler works tbh)

Got any more old versions around between 1.5.27-beta1 and 1.5.46 to keep trying? I think ultimately pinpointing the breaking code change would be very helpful, unless you already have some better clues on what's it

Thanks for your attention

**Comment by Rakashazi** [comment]
Ok, interesting results with Skip Late Frames On/Off as that may indicate some issue with the frame timestamps being reported by [Choreographer](https://developer.android.com/reference/android/view/Choreographer) which is used for timing on Android. I think the best way to tackle this now is keep testing old versions since coding up frame time stats will take more time.

Here's a set of 9 APKs to test out up to version 1.5.38 just before I switched to app bundles. Starting with 1.5.30 I added a native ARM64 build and switched to multi-APK so your device would have been served the sdk version 15 ones included.

https://drive.google.com/file/d/1sd3PAu0mzkrnYFjadme8M00H-YE_heuu/view

**Comment by Code-Otto** [comment]
Thanks for handing those APKs, I've been unable to test them though. Should have noted the important fact that despite this device featuring an AArch64 processor the entire system is running in Armv7 32-bit compatibility mode, possibly because of some hardware's proprietary blobs still unavailable for the Armv8-A architecture having forced the manufacturer to stick to a 32-bit kernel

Sorry for the inconvenience

**Comment by Rakashazi** [comment]
No problem, I guess we can rule out Armv7/Armv8 differences as a possible cause for the issue. I've updated the archive to have the sdk 9 versions with armv7 builds:

https://drive.google.com/file/d/1sd3PAu0mzkrnYFjadme8M00H-YE_heuu/view

**Comment by Code-Otto** [comment]
Out of the shared versions I've noticed 1.5.30 and onward seem to have higher CPU requirements as being ingame makes the frequency stay at its max of 1400MHz instead of around 1200MHz as with prior versions (guess a VBA-M core update?)

However none of them seem to be the culprit of this issue. Frame rate is good and similar in all versions from 1.5.28 to 1.5.38 and the emulators runs off the performance cores without manual hinting

Now the issue should hopefully appear somewhere between 1.5.38 and 1.5.46, sounds like we're closer at pinpointing it! :smiley:

**Comment by Rakashazi** [comment]
In 1.5.30 I added a hack in the system options called "Prevent CPU Idling" that generates fake touch events every sec to keep the CPU frequency high. It was removed in 1.5.35 since it only worked on specific Android versions and not even on all devices, but that's probably why the CPU freq increased.

It's on by default so turn it off manually since it'll mess with tracking down the original issue. Once those versions are re-tested here's hopefully the final batch of APKs. Since these ones come from app bundles I generated a universal APK for each release so they should run on any CPU type.

https://drive.google.com/file/d/1pKexKgSeh018Qi7xmKGL_Y4ajY6Xw2t_/view

**Comment by Code-Otto** [comment]
Re-tested through 1.5.30 to 1.5.34 with the 'Prevent CPU Idling' option disabled and I'm experiencing the same results, on 1.5.30 onward even past 1.5.34 where this option was removed CPU usage seems slightly higher

To better test this I've used my older Samsung Android 4.4.4 device. On it version 1.5.28 runs at full frame rate and even with 'Skip Late Frames' disabled there's no audio drops which indicates the emulation is keeping up fine
After updating to 1.5.30 under the same configuration (and independently of the 'Prevent CPU Idling' setting) a small audio drop is experienced around every second so if the emulation core has not been updated behind the scenes (can't seem to find a commit for it) there's been a small performance regression in the framework between 1.5.28 and 1.5.30

Let's keep looking forward for the big deal, however:
I spotted a new 'Render Multithreading' feature at 1.5.39, version which also brought in some severe performance regressions

**1.5.38**
First, a capture of 1.5.38 which still performs smoothly like 1.5.27-beta1 in the BQ device, for comparison

https://user-images.githubusercontent.com/129909770/232223000-240ae77a-1945-432d-85ca-e8a586e564de.mp4

**1.5.39**
Now here things get 'fun', performance experienced a heavy regression even with the new 'Render Multithreading' option disabled

Under heavy load (for example the weapon selection menu) almost all frames are dropped. Under normal load some heavy lag spikes are experienced too, notice them while I fight the mini-boss
Changing the 'GPU Copy Mode' and 'Skip Late Frames' appears to yield some improvements but frame rate is still far from good

https://user-images.githubusercontent.com/129909770/232223148-4d6d3593-eb20-43ed-bca5-e90e556fb87f.mp4

CPU usage is still high and the process is running off the performance cores, however interestingly in comparison to 1.5.38 frequencies go as low as 1200MHz again as with 1.5.27-beta1 though with bad performance instead
Using the fast-forward feature also causes almost all frames to drop like seen in the weapon selection menu

**1.5.43**
These terrible performance issues persist until version 1.5.43, where the menu becomes usable again
Notice the CPU core usage and the slowdowns with 'Skip Late Frames' disabled. This looks familiar!
This version introduces the main issue I initially opened this GitHub thread for

https://user-images.githubusercontent.com/129909770/232224439-5da7ae08-4731-4069-82eb-c9e04f447d74.mp4

1.5.43 and onward show this behavior... presumably all its way to the latest emulator version

Was really hoping the issue was some localized changed within two small releases, however this appears to be in between some heavy refactoring for the multithreaded rendered plus its later fixes. Now the bug looks more like Hydra from Greek mythology to me :cold_sweat:

**Comment by Rakashazi** [comment]
Thanks for all the testing so far. The initial renderer multithreading code defaulted to using two OpenGL contexts, one dedicated to uploading resources but it it definitely had some issues depending on the device's drivers.

Next let's use my FrameRateTest app to see how stable the frame times are so we can get a better idea of what's going on under the hood. Each test on the menu will show a flashing rectangle so it's easy to spot frame drops and it'll also change color if the app itself detects a frame drop. Ideally the total draw time should be fairly consistent like 2-3ms but not jump wildly. The timestamp diff should match your refresh rate exactly so 16ms for 60hz as an example. Here's a link to the APK:

https://drive.google.com/file/d/16H4y4jzfDdxDReuD0W8yen0alzPvdbPi/view

**Comment by Code-Otto** [comment]
Here's the test results:

- Clear:
   - **CPU:** ~8% 960MHz
   - **Total Draw Time:** 2-4ms (0ms)
   - **Timestamp Diff:** 16ms
   - **Detected frame rate**: 59.85 fps
- Draw RGB565 256x256 (System Memory):
   - **CPU:** ~8% 960MHz
   - **Total Draw Time:** 2-4ms (0ms)
   - **Timestamp Diff:** 16ms
   - **Detected frame rate**: 59.85 fps
 - Write RGB565 256x256 (System Memory):
   - **CPU:** ~8% 960MHz
   - **Total Draw Time:** 3-4ms (0ms)
   - **Timestamp Diff:** 16ms
   - **Detected frame rate**: 59.85 fps
- Draw RGB565 256x256 (OpenGL PBO):
   - **CPU:** ~8% 960MHz
   - **Total Draw Time:** 2-4ms (0ms)
   - **Timestamp Diff:** 16ms
   - **Detected frame rate**: 59.85 fps
- Write RGB565 256x256 (OpenGL PBO):
   - **CPU:** ~8% 960MHz
   - **Total Draw Time:** 3-4ms (0ms)
   - **Timestamp Diff:** 16ms
   - **Detected frame rate**: 59.85 fps
  
All tests appear to yield similar results aside from 'Draw' ones yielding a 'Total Draw Time' between 2-4ms and 'Write' ones 3-4ms

**Comment by Rakashazi** [comment]
It doesn't look like the core drawing code is the bottleneck so that's good to know. I went ahead and added some basic frame time stats to EmuFramework and built a new GBA.emu APK for you to test out. Currently the stats display is hardcoded but it'll be an option in a future update.

https://drive.google.com/file/d/1ZkX6EAaCUBZ7FB3UQUtcjQtAx6VY2Zt8/view

**Comment by Code-Otto** [comment]
Here's the results from this new version, iterating over combinations of the 'Skip Late Frames' on/off + 'No-Op thread' on/off options

https://user-images.githubusercontent.com/129909770/232463796-349e2f16-5db9-4051-a1af-bb0f482158ee.mp4

The options 'GPU Copy Mode', '* Color Format', 'Image Buffers' and 'Reduce Compositor Lag' didn't yield any notable changes to the timings as far as I have tested

That's some useful timing insights now... I see in the best case around ('Skip Late Frames' off + 'No-Op thread' on) my device can just barely reach <16ms frame processing times. Maybe I'm just asking way too much from it and this emulator...

**Comment by Rakashazi** [comment]
Ok, so it looks like the main bottleneck is indeed the CPU but the fact that the speed isn't ramping up without the no-op thread must mean something else is going on like unexpected blocking when calling an OS function. What's the average fps when using the benchmark function on Iridion 2? That just runs the first couple seconds of the game as fast as possible without calling the OS in between frames so we can see the raw CPU throughput. Also check if the CPU ramps up during the benchmark.

**Comment by Code-Otto** [comment]
The benchmark returns around 115fps on Iridion 2 with the CPU ramping up to its max frequency, and properly escalates from the low-power cores even without CPU affinity set up

P.S.: Gave VBA-M a try under RetroArch for comparison in the same device and it runs much slower there than with your frontend, even with this current performance bug. Congrats on how performant your frontend already is! :heart: I've always felt you put a lot of care into keeping your .EMU projects light and fast, and that's why I've always stick to them through the years
Another remarkable thing is I was greatly surprised by mGBA. It ran Iridion 2 at full speed without notable issues and so low CPU usage it can stay in the low-power cores just fine, even under RetroArch

If there's some more insightful statistics I could gather from the device, for example debug traces or profiling, I can give a try getting the Android SDK set up and capturing data. Please point out anything that could be of help

**Comment by Code-Otto** [comment]
Here's the results of testing with this version:
Looks like CPU affinity is not correctly being applied to the no-op thread in this version, so couldn't really get proper results for all the cases

https://user-images.githubusercontent.com/129909770/233109516-8a598b9c-bd89-40a5-99d4-487218a9c632.mp4

**Comment by Rakashazi** [comment]
Good catch with the no-op thread affinity, this new version only applies the custom affinity during emulation so the menu can still use the low powered cores, but as a result the no-op thread inherits it's affinity from the main thread before the affinity mask is applied.

Looking at the results I think I have a better idea of what's happening. Whenever a frame callback is missed because the previous frame is running late, the main thread will wait for the next callback to signal the emulation thread but this small delay causes the CPU clock to decrease which in turn makes it miss the next callback and this cycles continues. What I should try instead is have the main thread always update the emulation thread on what frame it should be on, even if it's running late so it can catch up asap. I'll post up a new APK once the implementation is ready and hopefully we won't even need the no-op thread anymore 🙂

**Comment by Rakashazi** [comment]
Here's a new APK to try. I didn't fix the no-op thread affinity since we want to test without it. Try with both stock affinity and big core affinity, and compare to the previous version without no-op thread. Hopefully now the emulator thread will stay busy on a missed frame callback and not cause a cascade of misses.

https://drive.google.com/file/d/1QC2PKIcxzRPxklUxBUmPFPslvbOi-UFT/view

**Comment by Code-Otto** [comment]
In this new version setting CPU affinity is no longer necessary, the process escalates into the big cores without any manual hinting again. Great! :smile: 

Should confess today I've not had proper sleep and feel so tired I can hardly engage in the already taxing task of evaluating framerates by eye but I do feel an improvement here

Find it interesting how the lowest rate of missed callbacks is obtained at the weapon selection menu, one of the most demanding parts in the game

Now 'Skip Late Frames' no longer makes a notable difference
Also, the frameskip feature appears to work much better now. At x8 it reaches a higher speed, even if my device can't clearly keep up, x2 feels more usable. I've however noticed disabling frameskip is not instantaneous anymore but keeps jamming for around a second.

https://user-images.githubusercontent.com/129909770/233375058-4520f596-4451-4b87-a8b4-32093720a77f.mp4

Here's what 8x frameskip looked like at the prior version instead for comparison:

https://user-images.githubusercontent.com/129909770/233373696-d7c8268c-9a08-49af-bbca-cf2c9617f7e6.mp4

**Comment by Rakashazi** [comment]
Nice! Looks like we found the main source of the issue so we can stick to simpler regression testing from now on. Thanks as always for all the reports. 🙂

As for the next version, I have an idea on how to fix the fast forward delay issue so check the next commit when it's up. We probably don't need the no-op thread any more so I'll compile it out of the builds but keep the code just in case it has some use in the future. I'm also going to add a new CPU affinity "mode" option with these settings:

- Auto: The default, will either use only the performance cores or try the the PerformanceHintManager API on newer versions of Android
- Any: Won't touch the affinity at all in case the user is setting it via an external method
- Manual: Will use the affinity configured in the menu as it's currently implemented

That should hopefully cover all the use cases and the reasoning to default to the performance cores even in cases where the emulation could run 100% on the smaller cores is that having lower frame times on Android allows the compositor to latch the finished frame earlier, possibly cutting out a frame of input lag.

**Comment by Code-Otto** [comment]
Been further testing version d995ef7 now with my senses back again operational and can conclude there's certainly been a frame rate improvement in my device, specially when option 'Reduce Compositor Lag' is disabled. It feels like 1.5.27-beta1 again!

Here's a comparison between 1.5.27-beta1, d995ef7 with 'Reduce Compositor Lag' enabled and same version again with such option disabled:

https://user-images.githubusercontent.com/129909770/233691327-35789fae-8ec0-4310-aaf2-df922ee02236.mp4

The difference is so subtle I wonder if I'm going insane... however there appears to be some eventual frame drops with said option enabled. For example, try closely looking in second 5 of the video at the smoothness of the screen frame transition in each version

As of the new CPU affinity settings scheme you plan on introducing, having this worked out by default for non-technical users sounds like a good idea

This is some good results now, looking forward for the frameskip fix :smile:

**Comment by Code-Otto** [comment]
Tested version 90df02a and the fast forward fix is doing good, it no longer jams and stops instantly upon user deactivation

The new CPU affinity menu looks and works great too! If something (though just nitpicky) it'd be nice to have the core list entries disabled while at the 'Auto' and 'Any' modes to tell the user the choices theres under said modes don't matter at all

I've tested this latest version under my older Samsung Android 4.4.4 device as well, as a more constrained test environment and emulation also performs pretty well there again, similar to how 1.5.30 did (just eventual frame skips)

As a small treat, here's what those low-power cores can actually do now:

https://user-images.githubusercontent.com/129909770/233811092-e407417a-c38b-467f-bee8-67bb4d94f9d7.mp4

That's pretty much playable! For turn-based strategy games (looking into Advance Wars) I plan on setting affinity to the low-power cores instead to save some battery time while at the train trips I play on
This bug that's just been fixed happened to make a great power-saving feature, if something. The emulator kept the CPU at its minimum which may fit useful for some use cases. Would it be possible to implement such a behavior in a less quirky way instead? For example a forced minimum of frame skip setting. Having the choice to reduce power consumption at the cost of frame rate sounds like an useful feature for the battery-powered devices this emulator mostly targets

I see another user around the issue tracker also reported this fixing long-standing performance issues for them, I'm glad this also turned helpful for more people :smiley: 
Thanks for all the attention here and the time put into fixing this! As of me, this issue can be marked closed now

**Comment by Rakashazi** [comment]
Awesome, thanks again for all your testing! 😀

If you have any Android 13 devices, give them a try in the auto mode since I implemented the new PerformanceHintManager API too. I tested it on a Pixel 7 and it works even better than setting affinities to just the big cores since it seems to boost frequencies and selects the cores that can reach the frame time deadline automatically. Overall it cut the frame times in half so it looks like thread scheduling problems will be mostly a non issue on newer devices going forward.

As for a minimum frameskip setting, I have something like it called "Target Frame Rate" on the iOS port since there's a simple OS function to slow down the vsync interval while keeping correct frame pacing. There isn't an exact equivalent on Android but I think it can be done by using Choreographer and tweaking the [presentation time extension](https://registry.khronos.org/EGL/extensions/ANDROID/EGL_ANDROID_presentation_time.txt) code a little so I'll look into it for a future update. Eventually I'd like to add simple preset to select if the user prefers low input latency vs higher latency but better power saving like in your case for playing a strategy game.

That should be everything for now so I'll go ahead and close the issue but feel free to post if anything related to this comes up in the future.
